core/arch/instr_shared.c

```diff
@@ -168,6 +168,8 @@ instr_init(dcontext_t *dcontext, instr_t *instr)
 void
 instr_free(dcontext_t *dcontext, instr_t *instr)
 {
+    if (instr_is_label(instr) && instr_get_label_callback(instr) != NULL)
+        (*instr->label_cb)(dcontext, instr);
```

Comment:
How about adding an assert that !TEST(INSTR_RAW_BITS_ALLOCATED, instr->flags) as a sanity check that we're not using both parts of the new union